balance_strategy.go

```diff
@@ -127,3 +155,904 @@ func balanceStrategyHashValue(vv ...string) uint32 {
+		for _, partition := range partitions {
+			partition2AllPotentialConsumers[topicPartitionAssignment{Topic: topic, Partition: partition}] = []string{}
+		}
+	}
+
+	// create a mapping of all consumers to all potential topic partitions that can be assigned to them
+	// also, populate the mapping of partitions to potential consumers
+	consumer2AllPotentialPartitions := make(map[string][]topicPartitionAssignment, len(members))
+	for memberID, meta := range members {
+		consumer2AllPotentialPartitions[memberID] = make([]topicPartitionAssignment, 0)
+		for _, topicSubscription := range meta.Topics {
+			// only evaluate topic subscriptions that are present in the supplied topics map
+			if _, found := topics[topicSubscription]; found {
+				for _, partition := range topics[topicSubscription] {
+					topicPartition := topicPartitionAssignment{Topic: topicSubscription, Partition: partition}
+					consumer2AllPotentialPartitions[memberID] = append(consumer2AllPotentialPartitions[memberID], topicPartition)
+					partition2AllPotentialConsumers[topicPartition] = append(partition2AllPotentialConsumers[topicPartition], memberID)
+				}
+			}
+		}
+
+		// add this consumer to currentAssignment (with an empty topic partition assignment) if it does not already exist
+		if _, exists := currentAssignment[memberID]; !exists {
+			currentAssignment[memberID] = make([]topicPartitionAssignment, 0)
+		}
+	}
+
+	// create a mapping of each partition to its current consumer, where possible
+	currentPartitionConsumer := make(map[topicPartitionAssignment]string, len(currentAssignment))
+	for memberID, partitions := range currentAssignment {
+		for _, partition := range partitions {
+			currentPartitionConsumer[partition] = memberID
+		}
+	}
+
+	// sort the topic partitions in order of priority for reassignment
+	sortedPartitions := sortPartitions(currentAssignment, prevAssignment, isFreshAssignment, partition2AllPotentialConsumers, consumer2AllPotentialPartitions)
+	unassignedPartitions := deepCopyPartitions(sortedPartitions)
+	for memberID, partitions := range currentAssignment {
+		// if a consumer that existed before (and had some partition assignments) is now removed, remove it from currentAssignment
+		if _, exists := members[memberID]; !exists {
+			for _, partition := range partitions {
+				delete(currentPartitionConsumer, partition)
+			}
+			delete(currentAssignment, memberID)
+			continue
+		}
+
+		// otherwise (the consumer still exists)
+		updatedPartitions := deepCopyPartitions(partitions)
+		for _, partition := range partitions {
+			if _, exists := partition2AllPotentialConsumers[partition]; !exists {
+				// if this topic partition of this consumer no longer exists remove it from currentAssignment of the consumer
+				updatedPartitions = removeTopicPartitionFromMemberAssignments(updatedPartitions, partition)
+				delete(currentPartitionConsumer, partition)
+			} else if _, exists := topics[partition.Topic]; !exists {
```

Comment:
I believe this conditional is different from the Java one, and that this conditional is not correct.

The Java conditional checks whether the member (for the member ID that is currently being iterated over) is _still_ asking for the topic that it did prior:
```java
!subscriptions.get(entry.getKey()).topics().contains(partition.topic())
```
"Does the topic set for the member in this join group request still contain the topic that this partition it was consuming prior is in?" i.e. is it still consuming what it was.

The Go conditional here, by using `topics`, is asking "Does the client itself still know of this partition's topic?"

That is the same as the prior case that checks `partition2AllPotentialConsumers`, since that map is initialized with all topic/partitions that exist in the client:
```go
 187         partition2AllPotentialConsumers := make(map[topicPartitionAssignment][]string)
 188         for topic, partitions := range topics {
 189                 for _, partition := range partitions {
 190                         partition2AllPotentialConsumers[topicPartitionAssignment{Topic: topic, Partition: partition}] = []string{}
 191                 }
 192         }
```

Reply:
You're totally right, there was zero test coverage that executed the block associated with this conditional. I've updated the expression to check for the currently-evaluated topic in the set of topics subscribed to by the consumer. I also added some test coverage for this specific condition.